Fix for hotkey manager memory leak

adirh3 opened the pull request:
Added potential fix for hotkey manager memory leak

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
Everytime a control with binded HotKey is created, a manager instance is created with reference to FinderNode which contains disposables that reference the control.
Even when the control is detached from logical tree these references keep the control alive causing memory leak.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
A control that uses the HotKey property (Button, MenuItem) will never get cleaned up by GC, causing managed memory leak.

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
Controls with Hotkey property will not be kept in memory after they are detached from logical tree.

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->
Make sure the hotkey manager instance disposes itself (I did not change its behavior at all)

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
<!--- List any breaking changes here. When the PR is merged please add an entry to https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes -->


## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

adirh3 committed:
Added potential fix for hotkey manager instances not clearing up after detaching from root

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4823) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4823) before we can accept your contribution.<br/><br/>:x: adirh3 [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4823)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=4823) it.</sub>

adirh3 committed:
Changed HotkeyManager.cs to stop/init based on detaching/attaching to visual tree

adirh3 committed:
Reverted HotkeyManager.cs
Changed Button.cs to remove it's HotKey property when detaching from visual tree

adirh3 committed:
Changed Hotkey reset to detatch from logical tree (was visual tree)

jmacato committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix

adirh3 committed:
Added docs explaining why need to change Hotkey on attach/detach from logical tree

Added the same logic to MenuItem.cs

Reverted auto indentation in Button.cs

adirh3 committed:
Merge remote-tracking branch 'origin/fixes/hotkey-manager-leak-fix' into fixes/hotkey-manager-leak-fix

adirh3 committed:
Reverted auto indentation in HotkeyManager.cs

adirh3 committed:
Added unit tests for enforcing manager releasing reference to the controls

Changed AncestorFinder.cs to dispose child nodes and subject (need review on this)

adirh3 committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix

adirh3 committed:
Added another GC collect to the HotKeyManagerTests.cs memory leak tests

Renamed tests

adirh3 commented:
@grokys  @danwalmsley 
This is ready for review, it fixes the hotkey manager leak we discussed.
Thanks

adirh3 committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix

adirh3 committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix

adirh3 committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix

jmacato commented on tests/Avalonia.Controls.UnitTests/Utils/HotKeyManagerTests.cs:
```diff
@@ -50,10 +55,119 @@ public void HotKeyManager_Should_Register_And_Unregister_Key_Binding()
+                lm.ExecuteLayoutPass();
+                
+                // The button should be collected since it's detached from the listbox
+                GC.Collect();
```
just curious, why are we forcing GC to do stuff here?

adirh3 replied:
The leak was that any control that used HotKey never got cleaned up by GC even after it detaches and is not referenced by anything. This test just makes sure that the hotkey manager does not keep references of the control after it detaches from logical tree, so I need to make the GC collect it to confirm that nothing reference the control anymore.

adirh3 committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix

danwalmsley commented on src/Avalonia.Controls/MenuItem.cs:
```diff
@@ -337,6 +338,9 @@ protected override IItemContainerGenerator CreateItemContainerGenerator()
         protected override void OnAttachedToLogicalTree(LogicalTreeAttachmentEventArgs e)
         {
+            if (_hotkey != null)
+                // Control attached again, set Hotkey to create a hotkey manager for this control
```
please use braces, even though its a single statement...

and put the comment next to the if () or above it.

adirh3 replied:
Added, thanks

danwalmsley commented on src/Avalonia.Controls/MenuItem.cs:
```diff
@@ -347,6 +351,13 @@ protected override void OnAttachedToLogicalTree(LogicalTreeAttachmentEventArgs e
         {
+            if (HotKey != null)
+            {
+                // This will cause the hotkey manager to dispose the observer and the reference to this control
```
move comment to if () * or above it.

adirh3 replied:
Done

danwalmsley commented on src/Avalonia.Controls/Button.cs:
```diff
@@ -217,6 +221,13 @@ protected override void OnAttachedToLogicalTree(LogicalTreeAttachmentEventArgs e
         {
+            if (HotKey != null)
+            {
+                // This will cause the hotkey manager to dispose the observer and the reference to this control
```
move the comment to next to the if ()

adirh3 replied:
Done

danwalmsley commented on src/Avalonia.Controls/Button.cs:
```diff
@@ -207,6 +208,9 @@ protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e
         protected override void OnAttachedToLogicalTree(LogicalTreeAttachmentEventArgs e)
         {
+            if (_hotkey != null)
+                // Control attached again, set Hotkey to create a hotkey manager for this control
```
use braces and move comment to next to the if ()

adirh3 replied:
Done

danwalmsley commented on src/Avalonia.Controls/Button.cs:
```diff
@@ -207,6 +208,9 @@ protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e
+            if (_hotkey != null)
+                // Control attached again, set Hotkey to create a hotkey manager for this control
+                HotKey = _hotkey;
             base.OnAttachedToLogicalTree(e);
```
Add a newline between

adirh3 replied:
Done

danwalmsley reviewed (dismissed):
nits.

adirh3 committed:
Changed if statements to have braces
Moved comments above if

adirh3 committed:
Merge remote-tracking branch 'origin/fixes/hotkey-manager-leak-fix' into fixes/hotkey-manager-leak-fix

MarchingCube committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix

adirh3 committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix

grokys committed:
Merge branch 'master' into pr/4823-fix-for-hotkey-manager-memory-leak

grokys commented on src/Avalonia.Controls/Button.cs:
```diff
@@ -207,6 +208,11 @@ protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e
         {
+            if (_hotkey != null) // Control attached again, set Hotkey to create a hotkey manager for this control
+            {
+                HotKey = _hotkey;
```
Not a huge fan of this because `HotKey` may have been set with a non-LocalValue priority and here it's being promoted to `LocalValue` priority. This probably won't be noticed most of the time because non-LocalValue values are usually set by styles which will be detached from the control when it leaves the logical tree, but I suspect it's likely to cause hard-to-diagnose problems down the line.

adirh3 replied:
@grokys Thanks for replying, if this changes anything - I have been using this code in my production app for about 3 months, didn't notice any problem with styles not applying (I guess this might happen when changing themes and my app supports that).

I understand that because both the HotkeyManager and the Control itself are managing the Hotkey it isn't as clean, but as far as I understand other solution would be to refactor the HotkeyManager and I understood you wouldn't want that.

Also, without this putting control that uses HotKey inside templates (like button inside ListBox) will cause **drastic** memory leak, and obviously I can't ship my app like that.
If there are any solutions you guys would accept, then I will gladly implement it.

grokys reviewed (changes requested):
Sorry it took so long to review this @adirh3 but I've not had much time for anything not related to the 0.10 release recently. Now that's out I'm going back over older PRs.

Anyway, this solution feels rather hackish and is likely to cause problems with styled value priorities, so I'm going to 👎 it for now, sorry. Though this fixes a real problem, it's likely to cause other subtle problems down the line.

grokys reviewed (approved):
Ok, yeah I think you're right @adirh3 - for the moment the problem with binding priorities is low priority compared to the actual leaks. It's kinda unlikely that people will be setting the hotkey using styles anyway.

In the medium term, yeah the hotkey stuff needs to be rewritten, but for now I'm going to give this a 👍 so we can at least avoid the leaks.

grokys committed:
Merge branch 'master' into fixes/hotkey-manager-leak-fix
